### ISSUE TITLE
No way to save/load model weights between PyTorch and LibTorch (C++) in a compatible way

### ISSUE THREAD

**Issue by FuryBaM** [pytorch/pytorch#150857]
### 🚀 The feature, motivation and pitch

LibTorch (C++) does not provide a way to save or load model weights in a way compatible with Python’s torch.save() / torch.load() or state_dict() / load_state_dict(). This makes it hard to share models between Python and C++ environments.

### Alternatives

I've managed to implement cross-compatibility between PyTorch (Python) and LibTorch (C++) by creating my own binary weight serializer. Currently, LibTorch lacks a native and simple way to save/load only model weights that are compatible with Python torch.nn.Module.state_dict() and vice versa.

Here is a working example of what I had to do manually (binary format):
**C++ Code (LibTorch)**
```cpp
void save_weights(const std::string& path) {
    std::ofstream file(path, std::ios::binary);
    to(torch::kCPU);
    for (const auto& param : parameters()) {
        auto data = param.data().contiguous();
        auto size = data.numel();
        file.write(reinterpret_cast<const char*>(&size), sizeof(size));
        file.write(reinterpret_cast<const char*>(data.data_ptr<float>()), size * sizeof(float));
    }
    file.close();
}

void load_weights(const std::string& path) {
    std::ifstream file(path, std::ios::binary);
    for (auto& param : parameters()) {
        int64_t size;
        file.read(reinterpret_cast<char*>(&size), sizeof(size));
        std::vector<float> param_data(size);
        file.read(reinterpret_cast<char*>(param_data.data()), size * sizeof(float));
        param.data().copy_(torch::tensor(param_data).view_as(param));
    }
    file.close();
}
```
**Python Code**
```python
def save_binary_weights(self, path):
    with open(path, "wb") as file:
        for param in self.parameters():
            data = param.data.contiguous().cpu().view(-1).numpy().astype("float32")
            size = len(data)
            file.write(struct.pack('Q', size))
            file.write(struct.pack(f'{size}f', *data))

def load_binary_weights(self, path: str):
    with open(path, "rb") as file:
        for param in self.parameters():
            size = struct.unpack('Q', file.read(8))[0]
            param_data = struct.unpack(f'{size}f', file.read(size * 4))
            param.data.copy_(torch.tensor(param_data).view_as(param))
```
**Why this matters**
LibTorch doesn’t support Python’s torch.save() and torch.load() directly due to incompatible serialization mechanisms. Also, TorchScript lacks the ability to export model weights easily for reloading in C++.

**It would be very helpful if PyTorch could support either of the following:**

- A standard binary format for model weights compatible between Python ↔ C++.
- A .load_state_dict() and .state_dict() equivalent in LibTorch.
- Built-in utility to read .pt or .pth weight-only files in C++.

This would massively simplify workflows for deploying or testing models across environments.

### Additional context

You see full code in the my repo [Toguzkumalak](https://github.com/FuryBaM/Toguzkumalak/tree/master/Toguzkumalak).
Results in libtorch c++
```
Policy: [0.110181, 0.122630, 0.111052, 0.103369, 0.105678, 0.114038, 0.108151, 0.115452, 0.109449]
Value prediction: 0.051766
State: [9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 0.000000, 0.000000, 0.000000]
```

in python
```python
if __name__ == "__main__":
    model = TNET()
    model.load_binary_weights("C:/Users/Akzhol/source/repos/Toguzkumalak/Toguzkumalak/build/Release/model_data/weights.dat")
    input_data = torch.tensor([9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 
                            9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 
                            9.000000, 9.000000, 0.000000, 0.000000, 0.000000], dtype=torch.float32).view(1, -1)
    model.eval()
    output = model(input_data)
    print(output)
```
```
(tensor([[0.1102, 0.1226, 0.1111, 0.1034, 0.1057, 0.1140, 0.1082, 0.1155, 0.1094]],
       grad_fn=<ExpBackward0>), tensor([[0.0518]], grad_fn=<TanhBackward0>))
```

**Comment by FuryBaM** [comment]
if you need names.
```cpp
    void save_weights(const std::string& path) {
        std::ofstream file(path, std::ios::binary);
        to(torch::kCPU);

        for (const auto& named_param : named_parameters()) {
            const auto& name = named_param.key();
            const auto& param = named_param.value();

            int64_t name_length = name.size();
            file.write(reinterpret_cast<char*>(&name_length), sizeof(name_length));
            file.write(name.c_str(), name_length);

            auto data = param.data().contiguous();
            auto size = data.numel();
            file.write(reinterpret_cast<char*>(&size), sizeof(size));
            file.write(reinterpret_cast<char*>(data.data_ptr<float>()), size * sizeof(float));
        }

        file.close();
    }

    void load_weights(const std::string& path) {
        std::ifstream file(path, std::ios::binary);

        for (auto& named_param : named_parameters()) {
            auto& param = named_param.value();

            int64_t name_length;
            file.read(reinterpret_cast<char*>(&name_length), sizeof(name_length));

            std::string name(name_length, ' ');
            file.read(&name[0], name_length);

            if (name != named_param.key()) {
                std::cerr << "Warning: Expected parameter name '" << named_param.key()
                    << "', but found '" << name << "'\n";
                continue;
            }

			printf("Loading parameter: %s\n", name.c_str());

            int64_t size;
            file.read(reinterpret_cast<char*>(&size), sizeof(size));

            std::vector<float> param_data(size);
            file.read(reinterpret_cast<char*>(param_data.data()), size * sizeof(float));

            param.data().copy_(torch::tensor(param_data).view_as(param));
        }

        file.close();
    }
```

**Comment by FuryBaM** [comment]
with buffers

    void save_weights(const std::string& path) {
        std::ofstream file(path, std::ios::binary);
        if (!file) throw std::runtime_error("Failed to open file for writing");

        for (const auto& pair : named_parameters()) {
            const std::string& name = pair.key();
            const torch::Tensor& param = pair.value();
            torch::Tensor tensor = param.cpu().contiguous();

            int64_t name_len = name.size();
            file.write(reinterpret_cast<const char*>(&name_len), sizeof(name_len));
            file.write(name.data(), name_len);

            c10::ScalarType dtype = tensor.scalar_type();
            file.write(reinterpret_cast<const char*>(&dtype), sizeof(dtype));

            auto shape = tensor.sizes();
            int64_t ndims = shape.size();
            file.write(reinterpret_cast<const char*>(&ndims), sizeof(ndims));
            for (auto dim : shape) {
                file.write(reinterpret_cast<const char*>(&dim), sizeof(dim));
            }

            int64_t num_elems = tensor.numel();
            file.write(reinterpret_cast<const char*>(tensor.data_ptr()), num_elems * tensor.element_size());
        }

        for (const auto& pair : named_buffers()) {
            const std::string& name = pair.key();
            const torch::Tensor& param = pair.value();
            torch::Tensor tensor = param.cpu().contiguous();

            int64_t name_len = name.size();
            file.write(reinterpret_cast<const char*>(&name_len), sizeof(name_len));
            file.write(name.data(), name_len);

            c10::ScalarType dtype = tensor.scalar_type();
            file.write(reinterpret_cast<const char*>(&dtype), sizeof(dtype));

            auto shape = tensor.sizes();
            int64_t ndims = shape.size();
            file.write(reinterpret_cast<const char*>(&ndims), sizeof(ndims));
            for (auto dim : shape) {
                file.write(reinterpret_cast<const char*>(&dim), sizeof(dim));
            }

            int64_t num_elems = tensor.numel();
            file.write(reinterpret_cast<const char*>(tensor.data_ptr()), num_elems * tensor.element_size());
        }
    }

    void load_weights(const std::string& path) {
        std::ifstream file(path, std::ios::binary);
        if (!file) throw std::runtime_error("Failed to open file for reading");

        std::unordered_map<std::string, torch::Tensor> params_map;
        for (auto& pair : named_parameters()) {
            params_map[pair.key()] = pair.value();
        }

        std::unordered_map<std::string, torch::Tensor> buffers_map;
        for (auto& pair : named_buffers()) {
            buffers_map[pair.key()] = pair.value();
        }

        while (file.peek() != EOF) {
            int64_t name_len;
            file.read(reinterpret_cast<char*>(&name_len), sizeof(name_len));
            std::string name(name_len, '\0');
            file.read(name.data(), name_len);

            torch::Tensor* target_tensor = nullptr;
            auto it_param = params_map.find(name);
            if (it_param != params_map.end()) {
                target_tensor = &it_param->second;
            }
            else {
                auto it_buf = buffers_map.find(name);
                if (it_buf != buffers_map.end()) {
                    target_tensor = &it_buf->second;
                }
                else {
                    throw std::runtime_error("Parameter or buffer " + name + " not found in model");
                }
            }

            torch::Tensor& param = *target_tensor;

            c10::ScalarType dtype;
            file.read(reinterpret_cast<char*>(&dtype), sizeof(dtype));
            if (dtype != param.scalar_type()) {
                throw std::runtime_error("Data type mismatch for " + name);
            }

            int64_t ndims;
            file.read(reinterpret_cast<char*>(&ndims), sizeof(ndims));
            std::vector<int64_t> shape(ndims);
            for (int i = 0; i < ndims; ++i) {
                file.read(reinterpret_cast<char*>(&shape[i]), sizeof(shape[i]));
            }

            auto expected_shape = param.sizes();
            if (shape != expected_shape) {
                throw std::runtime_error("Shape mismatch for " + name);
            }

            int64_t num_elems = 1;
            for (auto dim : shape) num_elems *= dim;
            file.read(reinterpret_cast<char*>(param.data_ptr()), num_elems * param.element_size());
        }
    }
